activesupport/lib/active_support/cache/redis_cache_store.rb

```diff
@@ -0,0 +1,389 @@
+      #
+      # Failsafe: Raises errors.
+      def decrement(name, amount = 1, options = nil)
+        instrument :decrement, name, amount: amount do
+          redis.decrby normalize_key(name, options), amount
+        end
+      end
+
+      # Cache Store API implementation.
+      #
+      # Removes expired entries. Not supported.
+      def cleanup(options = nil)
+        super
+      end
+
+      # Clear the entire cache on all Redis servers. Safe to use on
+      # shared servers if the cache is namespaced.
+      #
+      # Failsafe: Raises errors.
+      def clear(options = nil)
+        if namespace = merged_options(options)[namespace]
+          delete_matched "*", namespace: namespace
+        else
+          failsafe :clear do
+            redis.flushdb
+          end
+        end
+      end
+
+      private
+        # Store provider interface:
+        # Read an entry from the cache.
+        def read_entry(key, options = nil)
+          failsafe :read_entry do
+            deserialize_entry redis.get(key)
+          end
+        end
+
+        def read_multi_mget(*names)
+          options = names.extract_options!
+          options = merged_options(options)
+
+          keys_to_names = names.map { |name| [ normalize_key(name, options), name ] }.to_h
+          values = redis.mget(*keys_to_names.keys)
+
+          keys_to_names.zip(values).each_with_object({}) do |((key, name), value), results|
+            if value
+              entry = deserialize_entry(value)
+              unless entry.nil? || entry.expired? || entry.mismatched?(normalize_version(key, options))
+                results[name] = entry.value
+              end
+            end
+          end
+        end
+
+        # For compat with the wexpected options hash argument.
+        def write_entry(key, entry, options = nil)
+          write_entry_kwargs key, entry, **options
+        end
+
+        # Write an entry to the cache.
+        #
+        # Requires Redis 2.6.12+ for extended SET options.
+        def write_entry_kwargs(key, entry, unless_exist: false, raw: false, expires_in: nil, race_condition_ttl: nil, **options)
+          value = raw ? entry.value.to_s : serialize_entry(entry)
+
+          # If race condition TTL is in use, ensure that cache entries
+          # stick around a bit longer after they would have expired
+          # so we can purposefully serve stale entries.
+          if race_condition_ttl && expires_in && expires_in > 0 && !raw
+            expires_in += 5.minutes
+          end
+
+          modifiers = {}
+          modifiers[:nx] = unless_exist
+          modifiers[:px] = (1000 * expires_in.to_f).ceil if expires_in
+
+          failsafe :write_entry do
+            redis.set key, value, modifiers
+          end
+        end
+
+        # Delete an entry from the cache.
+        def delete_entry(key, options)
+          failsafe :delete_entry, returning: false do
+            redis.del key
+          end
+        end
```

Comment:
Note `returning: false` here. It's what `failsafe` returns when it rescues an exception. Defaults to `nil`, omitted elsewhere.